# 01 - Initial Data Inspection

This notebook is used to inspect the raw board game dataset before cleaning or modelling.

The goals of this notebook are:

- Load the raw dataset
- Check the number of rows and columns
- Inspect column names
- Check data types
- Check missing values
- Look for duplicate records
- Identify the prediction target and useful features

In [1]:
import pandas as pd
from pathlib import Path

# Check where the notebook is currently running from.
current_folder = Path.cwd()
print("Current folder:", current_folder)

# Because this notebook is inside the notebooks folder,
# we move one level up to reach the main project folder.
project_root = current_folder.parent

# Set the path to the raw dataset.
data_path = project_root / "data" / "raw" / "bgg_dataset.csv"

print("Dataset path:", data_path)
print("Dataset exists:", data_path.exists())

# Load the dataset into a pandas DataFrame.
df = pd.read_csv(data_path, sep=";", decimal=",")

# Show the number of rows and columns.
df.shape

Current folder: c:\Users\Zombie\Desktop\vscode-projects\board-game-rating-predictor\notebooks
Dataset path: c:\Users\Zombie\Desktop\vscode-projects\board-game-rating-predictor\data\raw\bgg_dataset.csv
Dataset exists: True


(20343, 14)

In [2]:
# Show the first five rows of the dataset.
df.head()

,ID,Name,Year Published,Min Players,Max Players,Play Time,Min Age,Users Rated,Rating Average,BGG Rank,Complexity Average,Owned Users,Mechanics,Domains
0,174430.0,Gloomhaven,2017.0,1,4,120,14,42055,8.79,1,3.86,68323.0,"Action Queue, Action Retrieval, Campaign / Bat...","Strategy Games, Thematic Games"
1,161936.0,Pandemic Legacy: Season 1,2015.0,2,4,60,13,41643,8.61,2,2.84,65294.0,"Action Points, Cooperative Game, Hand Manageme...","Strategy Games, Thematic Games"
2,224517.0,Brass: Birmingham,2018.0,2,4,120,14,19217,8.66,3,3.91,28785.0,"Hand Management, Income, Loans, Market, Networ...",Strategy Games
3,167791.0,Terraforming Mars,2016.0,1,5,120,12,64864,8.43,4,3.24,87099.0,"Card Drafting, Drafting, End Game Bonuses, Han...",Strategy Games
4,233078.0,Twilight Imperium: Fourth Edition,2017.0,3,6,480,14,13468,8.70,5,4.22,16831.0,"Action Drafting, Area Majority / Influence, Ar...","Strategy Games, Thematic Games"


## Inspect Columns and Data Types

In this section, we inspect the structure of the dataset.

We want to understand:

- What columns exist
- What type of data each column contains
- Which columns are numeric
- Which columns are text-based

In [3]:
# Show all column names.
df.columns

Index(['ID', 'Name', 'Year Published', 'Min Players', 'Max Players',
       'Play Time', 'Min Age', 'Users Rated', 'Rating Average', 'BGG Rank',
       'Complexity Average', 'Owned Users', 'Mechanics', 'Domains'],
      dtype='str')

In [4]:
# Show basic infromation about the dataset,
# including column names, non-null counts, and data types.
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 20343 entries, 0 to 20342
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  20327 non-null  float64
 1   Name                20343 non-null  str    
 2   Year Published      20342 non-null  float64
 3   Min Players         20343 non-null  int64  
 4   Max Players         20343 non-null  int64  
 5   Play Time           20343 non-null  int64  
 6   Min Age             20343 non-null  int64  
 7   Users Rated         20343 non-null  int64  
 8   Rating Average      20343 non-null  float64
 9   BGG Rank            20343 non-null  int64  
 10  Complexity Average  20343 non-null  float64
 11  Owned Users         20320 non-null  float64
 12  Mechanics           18745 non-null  str    
 13  Domains             10184 non-null  str    
dtypes: float64(5), int64(6), str(3)
memory usage: 3.6 MB


In [5]:
# Show the data type of every column in simple table.
df.dtypes

ID                    float64
Name                      str
Year Published        float64
Min Players             int64
Max Players             int64
Play Time               int64
Min Age                 int64
Users Rated             int64
Rating Average        float64
BGG Rank                int64
Complexity Average    float64
Owned Users           float64
Mechanics                 str
Domains                   str
dtype: object

In [6]:
# Show summary statistics for numeric columns.
df.describe()

,ID,Year Published,Min Players,Max Players,Play Time,Min Age,Users Rated,Rating Average,BGG Rank,Complexity Average,Owned Users
count,20327.000000,20342.000000,20343.000000,20343.000000,20343.000000,20343.000000,20343.000000,20343.000000,20343.000000,20343.000000,20320.000000
mean,108216.245142,1984.249877,2.019712,5.672221,91.294548,9.601485,840.971391,6.403227,10172.890429,1.991188,1408.457628
std,98682.097298,214.003181,0.690366,15.231376,545.447203,3.645458,3511.562220,0.935911,5872.831597,0.848903,5040.179315
min,1.000000,-3500.000000,0.000000,0.000000,0.000000,0.000000,30.000000,1.050000,1.000000,0.000000,0.000000
25%,11029.000000,2001.000000,2.000000,4.000000,30.000000,8.000000,55.000000,5.820000,5087.500000,1.330000,146.000000
50%,88931.000000,2011.000000,2.000000,4.000000,45.000000,10.000000,120.000000,6.430000,10173.000000,1.970000,309.000000
75%,192939.500000,2016.000000,2.000000,6.000000,90.000000,12.000000,385.000000,7.030000,15258.500000,2.540000,864.000000
max,331787.000000,2022.000000,10.000000,999.000000,60000.000000,25.000000,102214.000000,9.580000,20344.000000,5.000000,155312.000000


## Initial Observations

The dataset loads correctly and contains 20,343 rows and 14 columns.

The dataset includes a mix of numeric and text-based columns.

Numeric columns include:

- `ID`
- `Year Published`
- `Min Players`
- `Max Players`
- `Play Time`
- `Min Age`
- `Users Rated`
- `Rating Average`
- `BGG Rank`
- `Complexity Average`
- `Owned Users`

Text-based columns include:

- `Name`
- `Mechanics`
- `Domains`

Some colums contain missing values:

- `ID`
- `Year Published`
- `Owned Users`
- `Mechanics`
- `Domains`

Some numeric columns are also show values that may need further investigation during data cleaning:

- `Year Published` appears as a decimal value, such as `2017.0`
- `Year Published` has very old minimum values, including negative years
- `Max Players` has a very high maximum value
- `Play Time` has a very high maximum value
- `Min Age` includes values of `0`, which may represent missing or unclear age guidance

These issues are not fixed yet because this notebook is focused on understanding the raw dataset before cleaning.

## Check Missing Values

In this section, I'll check which columns contain missing values.

Missing values are important because they can affect mz analysis and machine learning models. Before cleaning the data, we first need to first understand where the missing values are and how many there is.

In [7]:
# Count missing values in each column.
df.isnull().sum()

ID                       16
Name                      0
Year Published            1
Min Players               0
Max Players               0
Play Time                 0
Min Age                   0
Users Rated               0
Rating Average            0
BGG Rank                  0
Complexity Average        0
Owned Users              23
Mechanics              1598
Domains               10159
dtype: int64

In [8]:
# Calculate missing values as percenteg
missing_percentege = (df.isnull().sum() / len(df)) * 100

missing_percentege

ID                     0.078651
Name                   0.000000
Year Published         0.004916
Min Players            0.000000
Max Players            0.000000
Play Time              0.000000
Min Age                0.000000
Users Rated            0.000000
Rating Average         0.000000
BGG Rank               0.000000
Complexity Average     0.000000
Owned Users            0.113061
Mechanics              7.855282
Domains               49.938554
dtype: float64

In [9]:
# Create a clerer missing values summary table
missing_summary = pd.DataFrame({
    "missing_count": df.isnull().sum(),
    "missing_percentege": (df.isnull().sum() / len(df)) * 100
})

# Show only columns that have missing values.
missing_summary[missing_summary["missing_count"] > 0]

,missing_count,missing_percentege
ID,16,0.078651
Year Published,1,0.004916
Owned Users,23,0.113061
Mechanics,1598,7.855282
Domains,10159,49.938554


### Missing Values Observations

The missing values check shows that 5 columns contain missing data.

Columns with missing values are:

- `ID`: 16 missing values, approximetely 0.08%
- `Year Published`: 1 missing value, approximetely 0.005%
- `Owned Users`: 23 missing values, approximetely 0.11%
- `Mechanics`: 1,598 missing values, approximetely 7.86%
- `Domains`: 10,159 missing values, approximetely 49.94%

The `ID`, `Year Published`, and `Owned Users` columns have very low levels of missing data.

The `Mechanics` column has a moderate amount of missing data.

The `Domains` column has a very high amount of missing data, with almost half of the values missing. This column may need special handling during the data cleaning and feature selection stages.

At this stage, the missing values are only being identified, not fixed.

## Check Duplicate Records

In this section, we check whether the dataset contains duplicate records.

Duplicate records matters because repeated rows can distort analysis and machine learning results.

In [10]:
# Count fully duplicated rows (check repetations)
duplicate_rows = df.duplicated().sum()

duplicate_rows

np.int64(0)

In [11]:
# Check whether any board game names appear more than once (versions/editions of same game)
duplicate_names = df["Name"].duplicated().sum()

duplicate_names

np.int64(367)

In [12]:
# Show examples of duplicated game names, if case any exists
df[df["Name"].duplicated(keep=False)].sort_values("Name").head(20)

,ID,Name,Year Published,Min Players,Max Players,Play Time,Min Age,Users Rated,Rating Average,BGG Rank,Complexity Average,Owned Users,Mechanics,Domains
13930,17183.0,1001,1900.0,2,4,60,12,31,6.43,13932,2.00,15.0,"Auction/Bidding, Trick-taking",NaN
14684,205495.0,1001,2016.0,2,8,45,8,55,6.11,14686,2.00,110.0,"Action Queue, Grid Movement, Simultaneous Acti...",NaN
9693,11284.0,1862,2000.0,4,7,360,12,46,7.35,9695,3.71,72.0,"Network and Route Building, Stock Holding, Til...",NaN
19215,4456.0,1862,1990.0,1,2,180,12,51,4.68,19217,2.38,170.0,"Dice Rolling, Events, Hexagon Grid, Simulation",Wargames
14751,274205.0,3D Labyrinth,2019.0,2,4,30,7,37,6.10,14753,1.50,82.0,"Modular Board, Network and Route Building, Poi...",NaN
19338,8308.0,3D Labyrinth,2002.0,2,4,30,5,138,5.13,19340,1.06,359.0,Modular Board,Children's Games
18553,67364.0,3D Tic Tac Toe,0.0,2,2,10,0,38,4.81,18555,1.00,53.0,"Pattern Building, Pattern Recognition",Abstract Games
19974,16387.0,3D Tic Tac Toe,1953.0,2,2,1,6,106,4.22,19976,1.25,160.0,"Pattern Building, Pattern Recognition, Set Col...",Abstract Games
8318,73312.0,4 Seasons,2010.0,2,2,15,10,153,6.33,8320,1.83,344.0,"Hand Management, I Cut You Choose",NaN
14291,213304.0,4 Seasons,2016.0,3,4,30,10,40,6.37,14293,2.00,183.0,"Betting and Bluffing, Hand Management, Trick-t...",NaN


### Duplicate Records Observations

The dataset contains 0 fully 100% duplicated rows.

There are 367 repeated occurrences in the `Name` column. However, repeated game names does not necessarily represent duplicate records. The examples show that games with the same name may have different IDs, publication (release) years, player counts, ratings, or mechanics.

These records may represent different editions, versions, or separate games that share the same title. Therefore, duplicated names should not be removed automatically and will need further review during data cleaning.

## Prediction Target and Candidate Features

The main prediction target for this project is:

* `Rating Average`

This means that the machine learning model will attempt to predict the average user rating of a board game.

Initial candidate features include:

* `Year Published`
* `Min Players`
* `Max Players`
* `Play Time`
* `Min Age`
* `Complexity Average`
* `Mechanics`
* `Domains`

Some columns will not initially be used as prediction features:

* `ID` is an identifier and does not describe the characteristics of a game.
* `Name` identifies the game but is not initially sutable as a direct numeric feature.
* `BGG Rank` is strongly connected to game ratings and could cause data leakage.
* `Users Rated` and `Owned Users` describe popularity after release and may not be available when predicting the rating of a new game.

The final feature list may change after further exploration, cleaning, and model testing.


In [13]:
# To define the value that the model will try to predict
target_column = "Rating Average"

# To define initial columns that may be used as model features
candidate_features = [
    "Year Published",
    "Min Players",
    "Max Players",
    "Play Time",
    "Min Age",
    "Complexity Average",
    "Mechanics",
    "Domains"
]

print("Prediction target:")
print(target_column)

print("\nCandidate features:")
for feature in candidate_features:
    print("-", feature)

Prediction target:
Rating Average

Candidate features:
- Year Published
- Min Players
- Max Players
- Play Time
- Min Age
- Complexity Average
- Mechanics
- Domains


In [14]:
# Select only numeric columns, then calculate their 
# correlation with prediction target
numeric_correlations = (
    df.select_dtypes(include="number")
    .corr()["Rating Average"]
    .sort_values(ascending=False)
)

numeric_correlations

Rating Average        1.000000
Complexity Average    0.480833
ID                    0.400619
Min Age               0.190299
Owned Users           0.177269
Users Rated           0.169652
Play Time             0.089887
Year Published        0.081232
Max Players          -0.039943
Min Players          -0.160354
BGG Rank             -0.739277
Name: Rating Average, dtype: float64

### Prediction Target and Correlation Observations

The main prediction target is `Rating Average`.

The initial candidate features describe characteristics of the board game, including publication (release) year, player counts, play time, minimum age, complexity, mechanics, and domains.

Among the numeric candidate features, `Complexity Average` has the strongest positive correlation with `Rating Average`, at approximately 0.48. This suggests a moderate relationship between game complexity and average user rating.

`Min Age`, `Play Time`, and `Year Published` have weaker positive correlations with the target. `Min Players` has a small negative correlation, while `Max Players` has very little linear relationship with the target.

`BGG Rank` has a strong negative correlation with `Rating Average`. However, it will not be used as a prediction feature because game rank is already strongly influenced by ratings.

`ID` also shows a positive correlation, but it is only an identifier and does not represent a any meaningful board game characteristic.

`Users Rated` and `Owned Users` describe popularity after release and will not initially be used when predicting ratings from a game's characteristics.

The text-based `Mechanics` and `Domains` columns do not appear in the numeric correlation results. They will require transformation before they can be used by a machine learning model.

Correlation measures only simple linear relationships and does not determine the final usefulness of a feature.


## Initial Data Understanding Summary

The raw BoardGameGeek dataset contains 20,343 rows and 14 columns.

The dataset includes numeric information about publication year, player cunts, play time, minimum age, user ratings, complexity, ownership, and ranking. It also includes text-based information describing game mechanics and domains.

### Data quality findings

Several issues were identified for further investigation during data cleaning:

* `ID` contains 16 missing values.
* `Year Published` contains 1 missing value.
* `Oned Users` contains 23 missing values.
* `Mechanics` contains 1,598 missing values, approximately 7.86% of the dataset.
* `Domains` contains 10,159 missing values, approximately 49.94% of the dataset.
* `Years Published` includes negative and unusually old values.
* `Max Player` contains very high values, including a maximum of 999.
* `Play Time` contains extreme values, including a maximum of 60,000 minutes.
* `Min Age` includes values of 0, which may represent missing or unclear age guidance.

The dataset contains no fully duplicated rows.

There are 367 repeated occurrences in the `Name` column. These should not be removed automatically because games with the same title may represent different editions, versions, publication years, or separate games sharing the same name.

### Prediction target

The initial prediction target is:

* `Rating Average`

This means the machine learning model will attempt to predict the average user rating of a board game.

### Initial candidate features

The initial candidate features are:

* `Year Published`
* `Min Players`
* `Max Players`
* `Play Time`
* `Min Age`
* `Complexity Average`
* `Mechanics`
* `Domains`

`Complexity Average` has the strongest positive correlation with the prediction target among the planned numeric features, at approximately 0.48.

`BGG Rank` will not be used as a prediction feature because it is already strongly influenced by ratings and could cause data leakage.

`ID` will not be used because it is only an identifier.

`Users Rsted` and `Owned Users` will not initially be used because they describe popularity after release and may not be available when predicting the rating of a new board game.

The final features lists may change after data cleaning, exploratory analysis, and model testing.
